### ISSUE TITLE
test: start offset in expected_region_highlight is off-by-one to the actual format of zsh's region_highlight

### ISSUE THREAD

**Issue by m0vie** [zsh-users/zsh-syntax-highlighting#759]
The actual format of `region_highlight` is different to the defined expectations `expected_region_highlight` in tests:

E.g. for a region describing the full buffer `BUFFER=abcd` zsh would use `0 4 default`, but the expectations in test cases need to define `1 4 default`.

This is extremely confusing:

```
## cursor-matchingbracket
# BUFFER=$': ((( )))'
ok 1 - [3,3] «(»
ok 2 - [4,4] «(»
ok 3 - [5,5] «(»
ok 4 - [7,7] «)»
not ok 5 - [8,8] «)» - expected (8 8 "default"), observed (8 8 "bracket-level-2").
ok 6 - [9,9] «)»
ok 7 - [9,9] «)»
ok 8 - cardinality check
# expected_region_highlight     region_highlight
# '3 3 bracket-level-1'         '2 3 bracket-level-1'
# '4 4 bracket-level-2'         '3 4 bracket-level-2'
# '5 5 bracket-level-3'         '4 5 bracket-level-3'
# '7 7 bracket-level-3'         '6 7 bracket-level-3'
# '8 8 default'                 '7 8 bracket-level-2'
# '9 9 bracket-level-1'         '8 9 bracket-level-1'
# '9 9 cursor-matchingbracket'  '8 9 cursor-matchingbracket'
```

We should either 
a) Convert everything to the test-specific format before test assertions
-> Advantage: the test format is more human-readable
-> Disadvantage: this is not the format that zsh actually uses
b) decrement all the start offsets in every test case by 1.

**Comment by danielshahaf** [comment]
>  a) Convert everything to the test-specific format before test assertions
> -> Advantage: the test format is more human-readable
> -> Disadvantage: this is not the format that zsh actually uses

The proposal wasn't very clear until I looked at the diff.  The problem with it is that the data being dumped is no longer the raw `$region_highlight` values computed by zsh, but a massaged version of them, and massaging can introduce or mask bugs.

>  b) decrement all the start offsets in every test case by 1.

Test expectations use a 1-based numbering system in order to make it easier to write and (manually) verify expectations: it's the same column numbering system that ${EDITOR}s and `cut(1)` use.

> The actual format of `region_highlight` is different to the defined expectations `expected_region_highlight` in tests:

Triage: No user-visible effect; no particular urgency; would be a flag day for all in-flight PRs involving test data; thus: not scheduling this for any particular milestone.

I recommend that you just manually verify `expected_region_highlight` against $BUFFER and trust the test harness to do the unit-conversion-aware comparison correctly.  I suppose we could also have the paste(1) output annotated with margin markers next to lines that failed comparison.

**Comment by danielshahaf** [comment]
The `expected (8 8 "default"), observed (8 8 "bracket-level-2")` part could be improved too.

**Comment by danielshahaf** [comment]
Further context: https://github.com/zsh-users/zsh-syntax-highlighting/issues/195#issuecomment-139814816

**Comment by m0vie** [comment]
The problem is really just half-open ranges vs inclusive ranges.

> The proposal wasn't very clear until I looked at the diff. The problem with it is that the data being dumped is no longer the raw `$region_highlight` values computed by zsh, but a massaged version of them, and massaging can introduce or mask bugs.

Exactly. But is this a problem? The expectations formulated in the test cases are not either. And if anyhting, it's most likely *that* place that is usually being looked at which could cause confusion.

IMHO having consistency everywhere is much more important than accurately representing zsh internals in one out of 4 places.

> The `expected (8 8 "default"), observed (8 8 "bracket-level-2")` part could be improved too.

How so? That's the only place where it's at least consistent.

How about simply changing the table headers when pasting the full tables?
instead of 
```
# expected_region_highlight     region_highlight
# '3 3 bracket-level-1'         '2 3 bracket-level-1'
...
```
just print
```
# expected regions              actual regions
# '3 3 bracket-level-1'         '3 3 bracket-level-1'
...
```

That way it should be clear enough that this is not necessarily the real zsh-internal format.

**Comment by danielshahaf** [comment]
> Exactly. But is this a problem? The expectations formulated in the test cases are not either.

I'd rather have the expectations easier to eyeball than the verification code, than vice versa.

> And if anyhting, it's most likely _that_ place that is usually being looked at which could cause confusion.

This hasn't caused confusion.

> IMHO having consistency everywhere is much more important than accurately representing zsh internals in one out of 4 places.

There are _five_ places.

We can't have a format that's both easy to eyeball and used throughout the system, so we should place the seam ("unit conversion") where it will least get in the way.

Remember that debugging something in an interactive shell with `zle -N f; f() typeset -p >/dev/tty region_highlight` is a thing.

See below.

> How so? That's the only place where it's at least consistent.

It lies about what was observed.

> How about simply changing the table headers when pasting the full tables?

Doesn't address already-stated concerns.

What might work is denoting the two different types of range differently and then showing _both_ the raw and postprocessed values. (Plus adding a marker on the left margin of lines that failed, per above)

**Comment by m0vie** [comment]
> I'd rather have the expectations easier to eyeball than the verification code, than vice versa.

True - The format we use in the test cases is easier to read. But as you say further down - it also lies about what is observed.

> This hasn't caused confusion.

It surely confused me, hence this issue and PR.
I was writing a test case based on what I observes in the *real* `region_highlight` (similar to your `zle -f` snipped above), copied that over to a test case and went from there. Then pondering about the e off-by-1 errors in the test failures and eventually (+1)-ing all the lower bounds.

> There are _five_ places.

Even more reason to opt for consistency


> What might work is denoting the two different types of range differently and then showing _both_ the raw and postprocessed values.

Makes things even more verbose and confusing :/


In my opinion the only overall consistent solution would be to use the real `region_highlight` format everywhere and to put up with the slightly harder maintenance of test-cases.

**Comment by danielshahaf** [comment]
>  I was writing a test case based on what I observes in the _real_ `region_highlight` (similar to your `zle -f` snipped above), copied that over to a test case and went from there. Then pondering about the e off-by-1 errors in the test failures and eventually (+1)-ing all the lower bounds.

That's a one-time-only learning curve.

The issue would not occur for anyone writing a test case using generate.zsh or by copying and modifying an existing test case.

It might be possible to address the issue by extending tests/README.md.  This approach has not been considered in the foregoing discussion.

> Makes things even more verbose and confusing :/

Are you suffering from NIH syndrome?

> In my opinion the only overall consistent solution would be to use the real `region_highlight` format everywhere and to put up with the slightly harder maintenance of test-cases.

Rejected.

**Comment by danielshahaf** [comment]
I've opened https://github.com/zsh-users/zsh-syntax-highlighting/issues/762 for the various individual improvements identified during this discussion.

As you have indicated you will accept no solution other than your own proposals, and as I have already indicated those designs are not acceptable, I am closing this issue and the associated PR.

Thanks for the bug report, the discussion, and helping those potential improvements be identified.
